src/UartSerial.cpp

```diff
@@ -104,22 +102,30 @@ void UartSerial::begin(uint32_t speed_, size_t config_)
 
 void UartSerial::end()
 {
-  if(_fd != -1){
-    flush();
-    // Restore the original settings
-    tcsetattr(_fd, TCSANOW, &_tio_original_config);
-    close(_fd);
-    _fd = -1;
-  }
+  if(_fd == -1) return;
+  
+  flush();
+
+  // Shut down the poll() thread
+  _polling = false;
+  _pollThread.join();
+
+  // Restore the original settings
+  tcsetattr(_fd, TCSANOW, &_tio_original_config);
+  close(_fd);
+  _fd = -1;
 }
 
 void UartSerial::flush()
 {
-  tcdrain(_fd);
+  if(_fd != -1)
+    tcdrain(_fd);
 }
 
 int UartSerial::read()
 {
+  if(_fd == -1) return -1;
+
   int buffer[1];
```

Comment:
The call would never read more than 1 byte from the file descriptor as determined by the 3rd argument. However to be explicit we can change the type to just a uint8_t.